packages/node_modules/@ciscospark/spark-core/src/interceptors/rate-limit.js

```diff
@@ -0,0 +1,128 @@
+
+// extracts the common identity API being called
+const idBrokerRegex = /.*idbroker.webex.com\/([^\/]+)/;
+
+/**
+ * @class
+ */
+export default class RateLimitInterceptor extends Interceptor {
+
+  /**
+   * @returns {RateLimitInterceptor}
+   */
+  static create() {
+    return new RateLimitInterceptor({spark: this});
+  }
+
+  /**
+  * constructor
+  * @param {mixed} args
+  * @returns {Exception}
+  */
+  constructor(...args) {
+    super(...args);
+    rateLimitExpiryTime.set(this, new Map());
+  }
+    /**
+   * @see {@link Interceptor#onRequest}
+   * @param {Object} options
+   * @returns {Object}
+   */
+  onRequest(options) {
+    if (this.isRateLimited(options.uri)) {
+      return Promise.reject(
+        new Error(`API rate limited ${options.uri}`)
+      );
+    }
+    return Promise.resolve(options);
+  }
+  /**
+   * @see {@link Interceptor#onResponseError}
+   * @param {Object} options
+   * @param {Error} reason
+   * @returns {Object}
+   */
+  onResponseError(options, reason) {
+    if (reason.statusCode === 429 && options.uri.includes(`idbroker`)) {
+      // set the retry after in the map
+      this.setRateLimitExpiry(options.uri, this.getRetryAfterTime(options));
+    }
+    return Promise.reject(reason);
+  }
+
+  /**
+   * @param {object} options associated with the request
+   * @returns {number} retry after time in seconds
+   */
+  getRetryAfterTime(options) {
+    const retryAfter = options.headers[`retry-after`] || null;
+    if (retryAfter === null || retryAfter <= 0) {
+      return 60;
+    }
+    else if (retryAfter > 3600) {
+      return 3600;
+    }
+    return retryAfter;
+  }
+
+  /**
+   * Sets the system time at which the rate limiting
+   * will expire in the rateLimitExpiryTime map.
+   * Assumes retryAfter is in seconds (S)
+   * @param {string} uri API issuing the rate limiting
+   * @param {number} retryAfter seconds until rate limiting expires
+   * @returns {bool} true is value was successfully set
+   */
+  setRateLimitExpiry(uri, retryAfter) {
+    const apiName = this.getApiName(uri);
+    const currTimeSec = new Date().getTime() / 1000;
+    const expiry = currTimeSec + retryAfter;
+    const dict = rateLimitExpiryTime.get(this);
+    return apiName && dict && dict.set(apiName, expiry);
+  }
+
+  /**
+   * returns true if the API is currently rate limited
+   * @param {string} uri
+   * @returns {Boolean} indicates whether or not the API is rate currently rate limited
+   */
+  getRateLimitStatus(uri) {
+    const apiName = this.getApiName(uri);
+    const currTimeSec = new Date().getTime() / 1000;
+    const dict = rateLimitExpiryTime.get(this);
+    return dict && dict.get(apiName) === null ? false : currTimeSec < dict.get(apiName);
+  }
+
+  /**
+   * splits the URI and returns the API name of idBroker
+   * @param {string} uri
+   * @returns {string}
+   */
+  getApiName(uri) {
+    const results = uri === null ? null : uri.match(idBrokerRegex);
```

Comment:
This is a bit complicated for a ternary

Reply:
This is the use case for ternary. I don't follow what is complicated here.

It has been verified with unit tests.

Reply:
I could skip this ternary and return null, but the second ternary is definitely necessary.

Reply:
There is code that prevents Keying nulls, so shouldn't be a concern.

The code being more compact with ternary is more readable imo, and since they're functionally equivalent,  is it really such a big deal to have ternary?

We're talking about the difference between grepping 7 or 8 lines of code if we use if statements versus 2 lines using ternary operators.

This strikes me as a coding style thing rather than a concern about functional correctness. Is there a style rule saying no ternary operators?

I could add a check in getRateLimitStatus that short circuits and returns if API name is null, but that would be pretty clunky and would obfuscate the method's function imo.  So at line 100 I could add:
if(apiName === null)
 return false;

However, since the expectation is the apiName returned is never null due to the fact we check the Regex against the URI before calling get rate limit status, you are now adding an extra null check on every single positive hit to any idBroker/identity calls. We would expect that apiName is never null, so optimizing with a short circuit for the unexpected case where it is null would be *less* efficient overall.

Reply:
Just to chime in here.

I think there is definitely some code style preference here, but also from an architecture standpoint there are some best practices we can use. So, sure there are checks in place higher up the stack that'll short circuit certain behavior, but I think if we're trying to ensure code isolation, then it's important to have checks in place for odd behaviors in case the code calling these Classes get funky. For example, in this implementation, I would think you'd want to have checks in `onResponseError` and `onRequest` for if `options` and `options.uri` are even defined.

I don't personally have an issue with ternaries, and I think if we actually put together a style guide, we'd explicitly encourage them where it makes sense, but compactness doesn't necessarily mean more understandable, and here it definitely took me some extra time to figure out what the ternary is doing.

Reply:
There is no check for options === null in any of the onError interceptor methods.

Provided there is error checking prior to inputting a value into the method at some stage, it should be fine. The most appropriate place to put the logic that does input validation is prior to adding to whatever data structure you want to use to store.

I don't see a legitimate use case for someone accessing the map outside of the accessors provided in the module. Any additional interactions needed should be added after examining the current assumptions and expectations with respect to the map. If people are adding methods, and messing with the map wily-nilly, no amount of preparation is going to prevent problems.

Additionally the module is reasonably small, so it shouldn't be hard for someone to grok the code  and make the appropriate changes.

The discussion around ternary operators is not one of compactness versus excessive catch statements per se; given that the work has been done, and the code is functionally correct, and we agree it is a style preference, I don't think it makes sense to write it again.

I could do an informal survey here at my site, and we'd probably find that most people prefer ternary operators, perhaps at your site, you don't find ternary as easy to understand as lengthy if-then statements and explicit null checks; I don't think one or another is necessarily right or wrong so much as it isn't necessary to reconstitute the solution.

Reply:
> given that the work has been done, and the code is functionally correct, and we agree it is a style preference

We don't agree it's *just* "a style preference". Two of the maintainers of this project have raised concerns about readability. We're not saying "don't use ternaries", but we are concerned that these particular ternaries are difficult to grok.

Reply:
Ternary operators can't get much simpler than this. From my perspective you are essentially dictating style.

I'll go ahead and change these, but if you don't want ternary operators being used, you need to have an official rule, and not ask me to change functionally correct code that is already done.